langchain_postgres/vectorstores.py

```diff
@@ -231,13 +293,16 @@ class PGVector(VectorStore):
       You will need to recreate the tables if you are using an existing database.
     * A Connection object has to be provided explicitly. Connections will not be
       picked up automatically based on env variables.
+    * langchain_postgres now accept async connections. If you want to use the async
+        version, you need to set `async_mode=True` when initializing the store or
+        use an async engine.
     """
 
     def __init__(
         self,
         embeddings: Embeddings,
         *,
-        connection: Optional[Connection] = None,
+        connection: Union[None, DBConnection, Engine, AsyncEngine, str] = None,
```

Comment:
Could we separate out the async and sync connections?

```python
connection: Union[None, DBConnection, Engine, str] = None,
connection_async: Union[ AsyncEngine, str] = None
```

And remove `async_mode` as a parameter

Reply:
@eyurtsev 

The `async_mode` parameter only needs to be removed if the user initializes the instance using a **URL**. If this is done from a connection or an engine, it is not necessary. This is automatically deduced from the instance type (sync or async). This parameter is *optional*.

This simplifies the migration of sync code to async to a simple rule.
With just one rule: _convert `xxx()` method calls to `async axxx()`.

This is why connection is of type `Union[None, DBConnection, Engine, AsyncEngine, str]`. I don't like the presence of two exclusive attributes. You have to test that if one is valued, the other isn't, and so on. It's not a simple interface.

All you have to do is give an instance, so that the class knows whether the mode is synchronous or asynchronous.

You can consider deleting `async_mode` if, and only if, you also delete the `str` type of connection. Otherwise, it is necessary. In fact, that's all it's good for. The code checks for consistency. An error is sent if an `AsyncConnection` is given with `use_async=False`, a strange situation.